src/gt4py/next/errors/exceptions.py

```diff
@@ -0,0 +1,111 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2023, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import textwrap
+from typing import Any, Optional
+
+from gt4py.eve import SourceLocation
+
+from . import formatting
+
+
+class CompilerError(Exception):
+    location: Optional[SourceLocation]
+
+    def __init__(self, location: Optional[SourceLocation], message: str):
+        self.location = location
+        super().__init__(message)
+
+    @property
+    def message(self) -> str:
+        return self.args[0]
+
+    def with_location(self, location: Optional[SourceLocation]) -> "CompilerError":
+        self.location = location
+        return self
+
+    def __str__(self):
+        if self.location:
+            loc_str = formatting.format_location(self.location, caret=True)
+            return f"{self.message}\n{textwrap.indent(loc_str, '  ')}"
+        return self.message
+
+
+class UnsupportedPythonFeatureError(CompilerError):
+    feature: str
+
+    def __init__(self, location: Optional[SourceLocation], feature: str):
+        super().__init__(location, f"unsupported Python syntax: '{feature}'")
+        self.feature = feature
+
+
+class UndefinedSymbolError(CompilerError):
+    sym_name: str
+
+    def __init__(self, location: Optional[SourceLocation], name: str):
+        super().__init__(location, f"name '{name}' is not defined")
+        self.sym_name = name
+
+
+class MissingAttributeError(CompilerError):
+    attr_name: str
+
+    def __init__(self, location: Optional[SourceLocation], attr_name: str):
+        super().__init__(location, f"object does not have attribute '{attr_name}'")
+        self.attr_name = attr_name
+
+
+class CompilerTypeError(CompilerError):
```

Comment:
Maybe `DSLTypeError`?  `GT4PySourceTypeError`? ...?

Reply:
Let's address this together with `CompilerError`.

Reply:
Renamed to `TypeError_`